### PR TITLE
Add rgbct and color_temperature light platforms

### PR DESCRIPTION
**Pull request by jesserockz**
# What does this implement/fix? 

A platform for where rgbww or cwww lights have a ct and brightness output instead of cool and warm channels.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1379

## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/color_temperature/light.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+CONF_CONSTANT_BRIGHTNESS = "constant_brightness"
+
+CONFIG_SCHEMA = cv.All(
+    light.RGB_LIGHT_SCHEMA.extend(
+        {
+            cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(CTLightOutput),
+            cv.Required(CONF_COLOR_TEMPERATURE): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_BRIGHTNESS): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_COLD_WHITE_COLOR_TEMPERATURE): cv.color_temperature,
+            cv.Required(CONF_WARM_WHITE_COLOR_TEMPERATURE): cv.color_temperature,
```
Comment (oxan) [inline esphome/components/color_temperature/light.py:24]: I think naming the options `min_color_temperature` / `max_color_temperature` is better here.

File: esphome/components/light/light_color_values.h [hunk 2]
```diff
@@ -181,6 +188,15 @@ class LightColorValues {
     }
   }
 
+  /// Convert these light color values to an CWWW representation with the given parameters.
+  void as_ct(float color_temperature_cw, float color_temperature_ww, float *color_temperature) const {
+    if (this->color_mode_ & ColorCapability::COLOR_TEMPERATURE) {
+      const float ct_level =
+          (this->color_temperature_ - color_temperature_cw) / (color_temperature_ww - color_temperature_cw);
+      *color_temperature = ct_level;
+    }
```
Comment (oxan) [inline esphome/components/light/light_color_values.h:201]: All the other `as_*` methods reset the pointer to zero if the capability isn't active, might be nice to do here as well for consistency.

File: esphome/components/rgbct/light.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    light.RGB_LIGHT_SCHEMA.extend(
+        {
+            cv.GenerateID(CONF_OUTPUT_ID): cv.declare_id(RGBCTLightOutput),
+            cv.Required(CONF_RED): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_GREEN): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_BLUE): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_COLOR_TEMPERATURE): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_WHITE_BRIGHTNESS): cv.use_id(output.FloatOutput),
+            cv.Required(CONF_COLD_WHITE_COLOR_TEMPERATURE): cv.color_temperature,
+            cv.Required(CONF_WARM_WHITE_COLOR_TEMPERATURE): cv.color_temperature,
```
Comment (oxan) [inline esphome/components/rgbct/light.py:32]: Same here with min/max color temperature naming.

File: esphome/components/rgbct/rgbct_light_output.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  void set_cold_white_temperature(float cold_white_temperature) { cold_white_temperature_ = cold_white_temperature; }
+  void set_warm_white_temperature(float warm_white_temperature) { warm_white_temperature_ = warm_white_temperature; }
+  void set_color_interlock(bool color_interlock) { color_interlock_ = color_interlock; }
+
+  light::LightTraits get_traits() override {
+    auto traits = light::LightTraits();
+    if (this->color_interlock_)
+      traits.set_supported_color_modes({light::ColorMode::RGB, light::ColorMode::COLOR_TEMPERATURE});
+    else
+      traits.set_supported_color_modes({light::ColorMode::RGB_COLOR_TEMPERATURE, light::ColorMode::COLOR_TEMPERATURE});
```
Comment (oxan) [inline esphome/components/rgbct/rgbct_light_output.h:29]: You don't need to support the `COLOR_TEMPERATURE` mode here as well. `RGB_COLOR_TEMPERATURE` is a superset of that mode, so if the `COLOR_TEMPERATURE` mode is requested it'll automatically be upconverted to `RGB_COLOR_TEMPERATURE`.

**Review by oxan — COMMENTED** [review]
Nice, I've wanted these platforms for a while but didn't get to it yet!

File: esphome/components/light/light_color_values.h (hunk 2 above)
Comment (jesserockz) [reply esphome/components/light/light_color_values.h:201]: But setting the color_temperature to 0 means changing the color temperature, not turning off the white (brightness to 0) which I originally had in here too and going to add back to deal with non-interlocked lighting

File: esphome/components/color_temperature/light.py (hunk 1 above)
Comment (jesserockz) [reply esphome/components/color_temperature/light.py:24]: This matches the rgbww/cwww light platform in terms of naming and I dont want to have to introduce a new key

Comment (oxan) [reply esphome/components/color_temperature/light.py:24]: True, but rgbww/cwww lights actually have a physical cold and warm white light with a fixed color temperature, while variable color temperature lights (such as this) don't.

File: esphome/components/light/light_color_values.h (hunk 2 above)
Comment (oxan) [reply esphome/components/light/light_color_values.h:201]: Oh yeah of course, you're right.

File: esphome/components/color_temperature/light.py (hunk 1 above)
Comment (jesserockz) [reply esphome/components/color_temperature/light.py:24]: Going to leave this as is as a use could choose kelvin in yaml and min/max would be the wrong way

File: esphome/components/light/light_color_values.h [hunk 5]
```diff
@@ -188,12 +188,23 @@ class LightColorValues {
...
+          (this->color_temperature_ - color_temperature_cw) / (color_temperature_ww - color_temperature_cw);
+      *color_temperature = ct_level;
+      const float brightness = gamma_correct(this->state_ * this->brightness_ * this->white_, gamma);
+      *white_brightness = brightness;
+    } else if (this->color_mode_ &
+               ColorCapability::COLOR_TEMPERATURE) {  // Handles lights in COLOUR_TEMPERATURE only only mode.
       const float ct_level =
           (this->color_temperature_ - color_temperature_cw) / (color_temperature_ww - color_temperature_cw);
       *color_temperature = ct_level;
+      this->as_brightness(white_brightness, gamma);
```
Comment (oxan) [inline esphome/components/light/light_color_values.h]: Maybe something like this is a bit cleaner (less code duplication)? `clang-format` will probably whine about how I formatted it on GitHub, though.

```suggestion
    const float white_level = this->color_mode_ & ColorCapability::RGB ? this->white_ : 1;
    if (this->color_mode_ & ColorCapability::COLOR_TEMPERATURE) {
      *color_temperature = 
          (this->color_temperature_ - color_temperature_cw) / (color_temperature_ww - color_temperature_cw);
      *white_brightness = gamma_correct(this->state_ * this->brightness_ * white_level, gamma);
```

**Review by oxan — APPROVED** [review]

File: esphome/components/rgbct/rgbct_light_output.h (hunk 4 above)
Comment (jesserockz) [reply esphome/components/rgbct/rgbct_light_output.h:29]: I exposed this so if a user does not set interlock, they can still operate in color_temp mode, with no color. This allows for the simple UI in HA with just the temp slider.

File: esphome/components/light/light_color_values.h (hunk 5 above)
Comment (jesserockz) [reply esphome/components/light/light_color_values.h]: I knew there would be a simpler way to do this...haha
Going to sleep now :zzz:

**Review by OttoWinter — APPROVED** [review]

**Pull request merged**